src/components/HybridAppMiddleware.tsx

```diff
@@ -1,104 +1,149 @@
+    authenticated: boolean;
     children: React.ReactNode;
 };
 
 type HybridAppMiddlewareContextType = {
-    navigateToExitUrl: (exitUrl: Route) => void;
     showSplashScreenOnNextStart: () => void;
 };
 const HybridAppMiddlewareContext = React.createContext<HybridAppMiddlewareContextType>({
-    navigateToExitUrl: () => {},
     showSplashScreenOnNextStart: () => {},
 });
 
 /*
  * HybridAppMiddleware is responsible for handling BootSplash visibility correctly.
  * It is crucial to make transitions between OldDot and NewDot look smooth.
+ * The middleware assumes that the entry point for HybridApp is the /transition route.
  */
-function HybridAppMiddleware(props: HybridAppMiddlewareProps) {
+function HybridAppMiddleware({children, authenticated}: HybridAppMiddlewareProps) {
     const {isSplashHidden, setIsSplashHidden} = useSplashScreen();
     const [startedTransition, setStartedTransition] = useState(false);
     const [finishedTransition, setFinishedTransition] = useState(false);
-    const navigation = useNavigation<StackNavigationProp<RootStackParamList>>();
+    const [forcedTransition, setForcedTransition] = useState(false);
 
-    /*
-     * Handles navigation during transition from OldDot. For ordinary NewDot app it is just pure navigation.
-     */
-    const navigateToExitUrl = useCallback((exitUrl: Route) => {
-        if (NativeModules.HybridAppModule) {
-            setStartedTransition(true);
-            Log.info(`[HybridApp] Started transition to ${exitUrl}`, true);
+    const initialURL = useContext(InitialURLContext);
+    const routeParams = useTransitionRouteParams();
+    const [exitTo, setExitTo] = useState<Route | HybridAppRoute | undefined>();
+
+    const [isAccountLoading] = useOnyx(ONYXKEYS.ACCOUNT, {selector: (account) => account?.isLoading ?? false});
+    const [sessionEmail] = useOnyx(ONYXKEYS.SESSION, {selector: (session) => session?.email});
+
+    const maxTimeoutRef = useRef<NodeJS.Timeout | null>(null);
+
+    // We need to ensure that the BootSplash is always hidden after a certain period.
+    useEffect(() => {
+        if (!NativeModules.HybridAppModule) {
+            return;
         }
 
-        Navigation.navigate(exitUrl);
+        maxTimeoutRef.current = setTimeout(() => {
+            Log.info('[HybridApp] Forcing transition due to unknown problem', true);
+            setStartedTransition(true);
+            setForcedTransition(true);
+            setExitTo(ROUTES.HOME);
+        }, CONST.HYBRID_APP_MAX_TRANSITION_TIMEOUT);
     }, []);
 
-    /**
-     * This function only affects iOS. If during a single app lifecycle we frequently transition from OldDot to NewDot,
-     * we need to artificially show the bootsplash because the app is only booted once.
-     */
+    // Save `exitTo` when we reach /transition route.
+    // `exitTo` should always exist during OldDot -> NewDot transitions.
+    useEffect(() => {
+        if (!NativeModules.HybridAppModule || !routeParams?.exitTo || exitTo) {
```

Comment:
I think we should clear the `exitTo` value under `Navigation.navigate(Navigation.parseHybridAppUrl(exitTo));` - it is possible that we will reenter NewDot with different `exitTo` value, and the hiding logic will not trigger

Reply:
Hmmm we clear it, but there is a separate method that handles such case. 
[showSplashScreenOnNextStart](https://github.com/software-mansion-labs/expensify-app-fork/blob/5132dd597e20e4fd79984bb3923b04ff6391e958/src/components/HybridAppMiddleware.tsx#L61)

Reply:
Drawback of this solution is that function [showSplashScreenOnNextStart](https://github.com/software-mansion-labs/expensify-app-fork/blob/5132dd597e20e4fd79984bb3923b04ff6391e958/src/components/HybridAppMiddleware.tsx#L61) comes from a hook. It works fine in current codebase, but in the future we might need to consider different approach

Reply:
I think that it may happen eg. when a user gets logged out from the app for some reason - there might be also another case, but not sure where 😄

Reply:
I verified limiations of current approach and introduced native events to track transitions.